### PR TITLE
[Operator] Add conv3d forward

### PR DESCRIPTION
**Pull request by IdleIdiot**
### Type of Change
New Feature

### Description
Implement the 3D convolution operator

### Issue
#312 

### Progress
- [ ] Change is properly reviewed (1 reviewer required, 2 recommended).
- [x] Change is responded to an issue.
- [x] Change is fully covered by a UT.

### Performance

image

accuracy
image

### CODE REVIEW

File: benchmark/core_shapes.yaml [hunk 1]
```diff
@@ -184,6 +184,17 @@ ConvBenchmark:
...
+Conv3DBenchmark:
+  shapes:
+    - [32, 64, 64, 64, 64, 32, 3, 3, 3, 1, 1, 1]
+    - [32, 64, 64, 64, 64, 32, 3, 3, 3, 1, 2, 1]
+    - [104, 16, 32, 32, 32, 32, 4, 4, 4, 1, 0, 1]
+    - [64, 32, 18, 18, 18, 32, 5, 5, 5, 2, 1, 1]
+    - [16, 32, 12, 12, 12, 24, 3, 3, 3, 2, 1, 1]
+    - [16, 32, 24, 24, 24, 24, 3, 3, 3, 1, 1, 2]
+    - [16, 32, 24, 24, 24, 24, 3, 3, 3, 2, 2, 2]
+    - [16, 32, 24, 24, 24, 24, 3, 3, 3, 1, 2, 2]
```
Comment (StrongSpoon) [inline benchmark/core_shapes.yaml]: generally we set 5 shapes for core mode ;)

File: benchmark/test_special_perf.py [hunk 2]
```diff
@@ -343,6 +343,56 @@ def conv2d_input_fn(shape, dtype, device):
...
+            "padding": padding,
+        },
+
+    torch.backends.cudnn.allow_tf32 = False
+    bench = Conv3DBenchmark(
+        input_fn=conv3d_input_fn,
+        op_name="conv3d",
+        torch_op=torch.nn.functional.conv3d,
+        dtypes=FLOAT_DTYPES,
+    )
```
Comment (StrongSpoon) [inline benchmark/test_special_perf.py:392]: Since `conv3d` is not yet registered in the ATEN library, we should explicitly set `gems_op` using `bench.set_gems`.

Comment (IdleIdiot) [reply benchmark/test_special_perf.py:392]: Thank you for your suggestion. After explicitly specifying the operator call, the performance issue was exposed. I am currently working hard to identify the root cause of the performance problem.
image

**Comment by IdleIdiot** [comment]
image
I have made some optimizations to the auto-tuning configuration, which has improved the operator's performance metrics to some extent. However, after trying various parameter combinations, the performance of the fp16 and bf16 cases still cannot achieve satisfactory results. Currently, I suspect that this might be related to the grid dividing the input data all at once for computation, where excessive thread block allocation leads to performance degradation. Next, I will try to add the upper limit of thread block allocation and simultaneously modify the operator code for verification.
Like this: 

```
grid = lambda META: (
    min(8192, triton.cdiv(in_n * out_depth * out_height * out_width, META["BLOCK_NI_DO_HO_WO"])),
    triton.cdiv(out_c // groups, META["BLOCK_CO"]),
    groups,
)

```
